packages/synapse-react-client/src/components/useImmutableTableQuery.ts

```diff
@@ -102,20 +130,17 @@ export default function useImmutableTableQuery(
     return request
   }, [initQueryRequestFromProps])
 
-  const [lastQueryRequest, setLastQueryRequestInternal] =
-    useState<QueryBundleRequest>(initQueryRequest)
-
-  function setLastQueryRequest(
-    newState: React.SetStateAction<QueryBundleRequest>,
-  ) {
-    setLastQueryRequestInternal(newState)
-    const request =
-      typeof newState === 'function' ? newState(lastQueryRequest) : newState
-
+  const {
+    committedState: currentQueryRequest,
+    uncommittedState: nextQueryRequest,
+    commit,
+    setUncommittedState: setNextQueryRequest,
+  } = useCommittedState<QueryBundleRequest>(initQueryRequest, queryRequest => {
     if (onQueryChange) {
-      onQueryChange(JSON.stringify(request.query))
+      const queryJsonString = JSON.stringify(queryRequest.query)
+      onQueryChange(queryJsonString)
     }
-  }
+  })
```

Comment:
Track two versions of state so we can easily roll back if the user cancels a filter change when we ask for confirmation